### PR TITLE
feat: Implement Toaster according to composition API

### LINKED ISSUE #27774: Implement Toaster component
Related to https://github.com/microsoft/fluentui/issues/26638

Fully implement the Toast component that behaves like the temporary one introduced in https://github.com/microsoft/fluentui/issues/27772

- [ ] aria live narration
- [ ] opt out of aria live narration

### PR DESCRIPTION
**Pull request by ling1726**
Implements the Toaster using the composition hook pattern. The Toaster component has one single root slot but it is mapped to potentially 4 different DOM elements (one for each toast position). The toaster does not support ref forwarding for this reason.

Fixes #27774

### CODE REVIEW

File: packages/react-components/react-toast/src/components/Toaster/Toaster.test.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  isConformant<ToasterProps>({
+    Component: Toaster,
+    displayName: 'Toaster',
+    requiredProps: { 'data-testid': testid } as ToasterProps,
+    getTargetElement: result => result.getByTestId(testid),
+    disabledTests: [
+      'component-has-root-ref',
+      'component-handles-ref',
+      'component-has-static-classnames-object',
+      'component-handles-classname',
```
Comment (ling1726) [inline packages/react-components/react-toast/src/components/Toaster/Toaster.test.tsx:18]: Imperative toast triggering means there's no easy way to render the slots that have the className - disabling for now but open to suggestions

File: packages/react-components/react-toast/src/state/useToaster.ts [hunk 2]
```diff
@@ -20,35 +20,26 @@ export function useToaster<TElement extends HTMLElement>(options: ToasterProps =
...
 
-  const getToastsToRender = React.useCallback(
-    <T>(cb: (position: ToastPosition, toasts: Toast[]) => T) => {
-      if (!toaster) {
-        return [];
-      }
-
-      const toRender = new Map<ToastPosition, Toast[]>();
-      const toasts = Array.from(toaster.toasts.values());
+  const toastsToRender = (() => {
```
Comment (ling1726) [inline packages/react-components/react-toast/src/state/useToaster.ts:23]: refactored this return to be better suited to our composition API

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/603293375d68e2288ef50990394f24b66e3d8250'>603293375d68e2288ef50990394f24b66e3d8250</a></sub>
<div id="bundleSizeComment9423"/>

**Review by bsunderhus — APPROVED** [review]

File: packages/react-components/react-toast/src/components/Toaster/renderToaster.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    <>
+      {renderAriaLive ? <AriaLive announceRef={announceRef} /> : null}
+      {hasToasts ? (
+        <Portal>
+          {slots.bottomLeft ? <slots.bottomLeft {...slotProps.bottomLeft} /> : null}
+          {slots.bottomRight ? <slots.bottomRight {...slotProps.bottomRight} /> : null}
+          {slots.topLeft ? <slots.topLeft {...slotProps.topLeft} /> : null}
+          {slots.topRight ? <slots.topRight {...slotProps.topRight} /> : null}
+        </Portal>
+      ) : null}
```
Comment (bsunderhus) [inline packages/react-components/react-toast/src/components/Toaster/renderToaster.tsx:30]:
```suggestion
      {hasToasts && (
        <Portal>
          {slots.bottomLeft ? <slots.bottomLeft {...slotProps.bottomLeft} /> : null}
          {slots.bottomRight ? <slots.bottomRight {...slotProps.bottomRight} /> : null}
          {slots.topLeft ? <slots.topLeft {...slotProps.topLeft} /> : null}
          {slots.topRight ? <slots.topRight {...slotProps.topRight} /> : null}
        </Portal>
      )}
```

nit

File: packages/react-components/react-toast/src/components/Toaster/useToaster.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    bottomLeft: resolveShorthand(rootProps, {
+      required: toastsToRender.has(TOAST_POSITIONS.bottomLeft),
+      defaultProps: {
+        children: toastsToRender.get(TOAST_POSITIONS.bottomLeft)?.map(toast => (
+          <Toast {...toast} announce={announce} key={toast.toastId} visible={isToastVisible(toast.toastId)}>
+            {toast.content as React.ReactNode}
+          </Toast>
+        )),
+        'data-toaster-position': TOAST_POSITIONS.bottomLeft,
+      } as ExtractSlotProps<Slot<'div'>>,
```
Comment (bsunderhus) [inline packages/react-components/react-toast/src/components/Toaster/useToaster.tsx:41]: nit: would be nice to add a comment to why this casting is required.

**Review by bsunderhus — APPROVED** [review]

File: packages/react-components/react-toast/src/components/Toaster/Toaster.types.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+   * There is no intention (currently) to let users customize the div for each toast position.
+   */
+  root: Slot<'div'>;
+};
+
+export type ToasterSlotsInternal = ToasterSlots & {
+  bottomRight?: Slot<'div'>;
+  bottomLeft?: Slot<'div'>;
+  topRight?: Slot<'div'>;
+  topLeft?: Slot<'div'>;
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toaster/Toaster.types.ts:17]: `*Start` and `*End*` instead?

File: packages/react-components/react-toast/src/components/Toaster/useToaster.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    topLeft: resolveShorthand(toastsToRender.has(TOAST_POSITIONS.topLeft) ? rootProps : null, {
+      defaultProps: {
+        children: toastsToRender.get(TOAST_POSITIONS.topLeft)?.map(toast => (
+          <Toast {...toast} announce={announce} key={toast.toastId} visible={isToastVisible(toast.toastId)}>
+            {toast.content as React.ReactNode}
+          </Toast>
+        )),
+        'data-toaster-position': TOAST_POSITIONS.topLeft,
+      } as ExtractSlotProps<Slot<'div'>>,
+    }),
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toaster/useToaster.tsx:62]: Can't it be a function? I.e. `createToastSlot()` or something like that?

File: packages/react-components/react-toast/src/components/Toaster/useToasterStyles.styles.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+export const useToasterStyles_unstable = (state: ToasterState): ToasterState => {
+  const styles = useStyles();
+  if (state.bottomLeft) {
+    state.bottomLeft.className = mergeClasses(toasterClassNames.root, styles.root, state.root.className);
+    state.bottomLeft.style ??= {};
+    Object.assign(state.bottomLeft.style, getPositionStyles(TOAST_POSITIONS.bottomLeft, state.offset));
+  }
+
+  if (state.bottomRight) {
+    state.bottomRight.className = mergeClasses(toasterClassNames.root, styles.root, state.root.className);
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/react-components/react-toast/src/components/Toaster/useToasterStyles.styles.ts:33]: actually you're right

**Review by layershifter — APPROVED** [review]

**Pull request merged**